### PR TITLE
PIX: split vector alloca writes to enable debug instrumentation

### PR DESCRIPTION
**Pull request by jeffnn**
Bit of background: PIX's debug passes add new allocas, stores to which tie debug info to DXIL Values. 
In the case of a preexisting alloca, PIX will still add its debug writes, but the codegen may have emitted vector-valued stores. Concretely, this happens for the ray payload and RayDesc structs in DXR. Those vector-values stores were being treated incorrectly, resulting in missing values in the PIX shader debugger.

This change splits vector-valued stores to such allocas into extractelement/scalar-stores, which enables the rest of the PIX instrumentation to function as expected. 

(Worth noting that this change only applies to the PIX shader debugging pass.)

### CODE REVIEW

**Review by damyanp — COMMENTED** [review]
LGTM, but we should get a review from someone more familiar with all of this. @adam-yang perhaps?

Is there a reason that the testing for this isn't filecheck based?

File: lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp [hunk 1]
```diff
@@ -133,6 +134,18 @@ bool DxilAnnotateWithVirtualRegister::runOnModule(llvm::Module &M) {
   auto instrumentableFunctions =
       PIXPassHelpers::GetAllInstrumentableFunctions(*m_DM);
 
+  // Make a vector of the instructions since we may delete some of them as we go
```
Comment (adam-yang) [inline lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp:137]: There's a way to iterate safely while deleting:
```
  for (auto *F : instrumentableFunctions) {
    for (auto &block : F->getBasicBlockList()) {
      for (auto it = block.begin(); it != block.end();) {
        llvm::Instruction &I = *(it++);
        SplitVectorStores(m_DM->GetOP(), I);
      }
    }
  }
```

File: lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp [hunk 2]
```diff
@@ -433,6 +474,39 @@ void DxilAnnotateWithVirtualRegister::AssignNewAllocaRegister(
...
+
+  llvm::AllocaInst *Alloca;
+  llvm::Value *Index;
+  if (!IsAllocaRegisterWrite(pSt->getPointerOperand(), &Alloca, &Index)) {
+    return;
+  }
+
+  llvm::Type *SourceType = pSt->getValueOperand()->getType();
+  const llvm::Type::TypeID ID = SourceType->getTypeID();
+  if (ID == llvm::Type::TypeID::VectorTyID) {
```
Comment (adam-yang) [inline lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp:492]: Can just call `isVectorTy()`

File: lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp [hunk 3]
```diff
@@ -433,6 +474,39 @@ void DxilAnnotateWithVirtualRegister::AssignNewAllocaRegister(
...
+    if (auto *constIntIIndex = llvm::cast<llvm::ConstantInt>(Index)) {
+      // break vector alloca stores up into individual stores
+      llvm::IRBuilder<> B(pSt);
+      auto *source = llvm::dyn_cast<llvm::VectorType>(SourceType);
```
Comment (adam-yang) [inline lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp:496]: No need to dyn_cast it, just call `SourceType->getVectorNumElements()`. Gets rid of the shadowing of the name `source` too.

File: lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp [hunk 4]
```diff
@@ -297,15 +310,41 @@ bool DxilAnnotateWithVirtualRegister::IsAllocaRegisterWrite(
...
+
+        // And the source pointer may be a vector (floatn) type,
+        // and if so, that's another offset to consider.
+        llvm::Type *DestType = pGEP->getPointerOperand()->getType();
+        const llvm::Type::TypeID ID = DestType->getTypeID();
+        // We expect this to be a pointer type (it's a GEP after all):
+        if (ID == llvm::Type::TypeID::PointerTyID) {
+          llvm::Type *PointedType =
+              llvm::cast<llvm::PointerType>(DestType)->getElementType();
+          const llvm::Type::TypeID PointedID = PointedType->getTypeID();
```
Comment (adam-yang) [inline lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp:334]: Can just be `PointedType->isPointerTy()`.

**Review by adam-yang — CHANGES_REQUESTED** [review]
Only big one is the leak.

File: lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp (hunk 1 above)
Comment (jeffnn) [reply lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp:137]: Thanks. I guess both approaches implicitly rely on the fact that the container doesn't reallocate elements...

File: lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp [hunk 5]
```diff
@@ -433,6 +474,39 @@ void DxilAnnotateWithVirtualRegister::AssignNewAllocaRegister(
...
+      llvm::IRBuilder<> B(pSt);
+      auto *source = llvm::dyn_cast<llvm::VectorType>(SourceType);
+      for (uint64_t el = 0; el < source->getVectorNumElements(); ++el) {
+        llvm::Value *destPointer = B.CreateGEP(pSt->getPointerOperand(),
+                                               {B.getInt32(0), B.getInt32(el)});
+        llvm::Value *source =
+            B.CreateExtractElement(pSt->getValueOperand(), el);
+        B.CreateStore(source, destPointer);
+      }
+      pI->removeFromParent();
```
[the remark this answers is not in the document]
Comment (jeffnn) [reply lib/DxilPIXPasses/DxilAnnotateWithVirtualRegister.cpp:504]: Thanks!

**Comment by jeffnn** [comment]
> LGTM, but we should get a review from someone more familiar with all of this. @adam-yang perhaps?
> 
> Is there a reason that the testing for this isn't filecheck based?

Yes: you can't get the textual pass "printf" output via filecheck.

**Review by damyanp — APPROVED** [review]

**Comment by damyanp** [comment]
Still LGTM - changing my review into an approval!

**Pull request merged**